receiver/prometheusreceiver/internal/staleness_store.go

```diff
@@ -0,0 +1,78 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+type stalenessStore struct {
+	currentHashes  map[uint64]bool
+	previousHashes map[uint64]bool
+	previous       []labels.Labels
+	current        []labels.Labels
+}
+
+func newStalenessStore() *stalenessStore {
+	return &stalenessStore{
+		previousHashes: make(map[uint64]bool),
+		currentHashes:  make(map[uint64]bool),
+	}
+}
+
+// refresh copies over all the current values to previous, and prepares.
+func (ss *stalenessStore) refresh() {
+	// 1. Clear ss.previousHashes firstly. Please don't edit
+	// this map clearing idiom as it ensures speed.
+	for hash := range ss.previousHashes {
+		delete(ss.previousHashes, hash)
+	}
+	// 2. Copy over ss.currentHashes to ss.previousHashes.
+	for hash := range ss.currentHashes {
+		ss.previousHashes[hash] = ss.currentHashes[hash]
+	}
+	// 3. Clear ss.currentHashes, with the map clearing idiom for speed.
+	for hash := range ss.currentHashes {
+		delete(ss.currentHashes, hash)
+	}
+	// 4. Copy all the prior labels from what was previously ss.current.
+	ss.previous = ss.current
+	// 5. Clear ss.current to make for another cycle.
+	ss.current = nil
+}
+
+// isStale returns whether lbl was seen only in the previous scrape and not the current.
+func (ss *stalenessStore) isStale(lbl labels.Labels) bool {
+	hash := lbl.Hash()
+	return ss.previousHashes[hash] && !ss.currentHashes[hash]
+}
+
+// markAsCurrentlySeen adds lbl to the manifest of labels seen in the current scrape.
```

Comment:
Maybe add a comment saying that this method needs to be called before refresh if a scrape has seen a label.

Reply:
Done, thanks!